### DISCUSSION TITLE
Did I say "thank you" already?

### DISCUSSION THREAD

**Discussion by johnwoltman** [fooyin/fooyin#230] [General]
If not, thank you!  I've been experimenting with layouts and I feel it's so much easier than customizing Foobar.  I particularly like how, when you're in editing mode, each widget has a "split" option and a "move" option.  It makes moving things around very simple.  Thanks again for this great software.

**Comment by ludouzi** [comment]
Thanks! One of the goals was to make layout editing as easy as possible, so I'm glad to hear that you're finding it straightforward. There's still a ways to go until a stable 1.0 release, but I'm glad it's useful to so many people already.

**Comment by misecichin** [comment]
Thank you! A true hero! The only piece of software which I missed from Windows.

**Comment by mattx1986** [comment]
Thank You for this great piece of software! The editing layouts is really pleasant. I created my own folder browser layout and I'm very happy with how fast and pleasant to use Fooyin is. Qt with Breeze theme looks fine on GTK based MATE desktop, but I hope for GTK support in the future similar to Audacious that supports both Qt and GTK with switch in the settings. Fooyin may become the best music player on Linux along with Audacious.

**Comment by OrangeRollo** [comment]
A big +1 here. Recent versions have been fantastic and each new build is better than the last.

**Comment by bjlanger** [comment]
This is awesome! I've been looking for a foobar2000 replacement and nothing out there was cutting it.

**Comment by ghost** [comment]
Adding my huge "Thank You!". I've been trying to drop Windows for a long time, but the biggest roadblock is lack of software alternatives on Linux and Music Bee has been my favorite for years and one of the top apps I'd be missing if I switched to Linux. I know I'm not the only one in this boat as there is not a single Linux music app that is even close to MusicBee, until now :)

I was able to achieve a Fooyin layout that is very close to what I have now in MusicBee. There are some quirks but they're minor and I did have some experience with Foobar 2000 so this felt somewhat familiar, still there is some learning curve for sure.

It looks very good in KDE Plasma too.

All I need now is Smart Playlists and one major roadblock on the road to Linux will be gone :)

Thank you very much and looking forward to the next update!

**Comment by ajtompkins** [comment]
Ditto! Thanks for all the hard work that's been put into this! 

I tried like a million different music players after switching to Linux and nothing scratched that foobar itch. I ended up running a slightly wonky install of Foobar via Wine. Then I found fooyin. Happy days.

**Comment by psychichlawyer** [comment]
A huge thank you for all the work that went into this fine music player, it is highly appreciated! ❤️

**Comment by droidwayin** [comment]
My huge gratitude for the devs behind this project. I mostly have FLAC music collection and tried many audio players but Fooyin is the superior player in terms of audio quality. I didn't had a github account and I only created one to make posts in this project. Thank you so much for building it. LoVe ❤️

**Comment by wesleywerner** [comment]
Another BIG thank you to everyone behind this project, it really is remarkable and I am so happy to have found it as a Foobar alternative. Thank You!

Aside: I wrote a Python script to migrate my playback statistics from Foobar2000 to fooyin. Perhaps someone else may find it useful.

```python
# A Python script to migrate foobar2000 playback statistics to fooyin.
# This requires that you used the foobar plugin called "foo_playcount"
# to track your playback stats.
# Use the Library > Playback Statistics > Export menu to dump an XML
# then run this script to generate the SQL update script to migrate.

# Important: Fill in the path variables below
#
# XML_INPUT_PATH: the location of the XML to process, you know the one
#                 you exported from Foobar.
#
# WINDOWS_LIBRARY_PATH: the ROOT location where your files lived on your
#                       Windows machine, open the exported XML
#                       to see it, e.g. "S:\music" is the root of all music.

# usage: python fooyin_migrate_fb2k.py
# outputs: fooyin_playback_stats.sql

import os
import xml.etree.ElementTree as ET
from pathlib import PureWindowsPath, PurePosixPath

XML_INPUT_PATH = "~/migrate/foobar2k_playback_stats_2025.xml"
WINDOWS_LIBRARY_PATH = r"S:\music"

def convert_to_unix_epoch_milliseconds(friendly_date):
    """Convert a friendly date string to Unix epoch in milliseconds."""
    from datetime import datetime
    dt = datetime.strptime(friendly_date, '%Y-%m-%d %H:%M:%S')
    return int(dt.timestamp() * 1000)

def generate_sql_statements(xml_file, output_file):
    global WINDOWS_LIBRARY_PATH
    # Ensure windows path is terminated correctly
    if not WINDOWS_LIBRARY_PATH[:-1] == "\\":
        WINDOWS_LIBRARY_PATH = WINDOWS_LIBRARY_PATH + "\\"
    # Parse the XML file
    tree = ET.parse(xml_file)
    root = tree.getroot()

    with open(output_file, 'w') as sql_file:
        for entry in root.findall('Entry'):
            count = int(entry.get('Count', 0))
            if count == 0:
                continue  # Skip entries with Count = 0

            # Extract values from the entry
            first_played_friendly = entry.get('FirstPlayedFriendly')
            last_played_friendly = entry.get('LastPlayedFriendly')
            rating = entry.get('RatingFriendly')
            entry_item = entry.find('Item')
            entry_id = entry.get('ID')
            if entry_item is None:
                print(f"- entry id {entry_id} missing Item")
                continue
            item_path = entry_item.get('Path')
            
            # convert windows path to linux path
            windows_path = PureWindowsPath(item_path.replace(WINDOWS_LIBRARY_PATH, ""))
            posix_path = PurePosixPath(*windows_path.parts)
            item_linux_path = str(posix_path)
            
            # escape quotes for sql
            item_linux_path = item_linux_path.replace("'", "''")

            # Convert friendly dates to Unix epoch in milliseconds
            first_played_epoch = convert_to_unix_epoch_milliseconds(first_played_friendly)
            last_played_epoch = convert_to_unix_epoch_milliseconds(last_played_friendly)
            
            # Conditionally updating rating.
            # Note we need to transform rating to a value fooyin understands.
            rating_update = ""
            if not rating is None:
                if rating == "1":
                    rating = "0.200000002980232"
                elif rating == "2":
                    rating = "0.400000005960464"
                elif rating == "3":
                    rating = "0.600000023841858"
                elif rating == "4":
                    rating = "0.800000011920929"
                elif rating == "5":
                    rating = "1"
                else:
                    print(f"entry id {entry_id} has rating of {rating} but I don't know how to map it.")
                    rating = None
            
            if not rating is None:
                rating_update = f"Rating = {rating}, "

            sql_statement = (
                f"UPDATE TrackStats "
                f"SET FirstPlayed = {first_played_epoch}, "
                f"LastPlayed = {last_played_epoch}, "
                f"{rating_update}"
                f"PlayCount = {count} "
                f"WHERE TrackHash IN ( "
                f"SELECT TrackHash "
                f"FROM Tracks "
                f"WHERE FilePath LIKE '%{item_linux_path}'"
                f");\n"
                );
            
            # Write the SQL statement to the file
            sql_file.write(sql_statement)

generate_sql_statements(os.path.expanduser(XML_INPUT_PATH), 'fooyin_playback_stats.sql')
```

**Comment by git-FaV** [comment]
Hello everyone,

I made an account just to be able to chip in here. fooyin is starting to look really great. foobar2000 is the single most useful software that I have yet to find a proper Linux alternative for.

Thank you!

@wesleywerner Thanks to you, too. You should really think about making this into a plugin, where paths can be chosen and database import is handled automatically. I lack the skills, unfortunately. Your script worked, though.

Kind regards

**Comment by tbonanno** [comment]
Just wanted to chime in and voice my support too. Foobar2000 was the gold standard of media players, and none of the popular Linux alternatives even came close to meeting the mark. Fooyin shows a lot of promise in being nearly as customizable as foobar2000, all while having a more slick default UI, adhering to my KDE menus and colors. Just installed it and spent an hour or so customizing my UI, and I'm already fully committed. Can't wait to see how development goes in the future, and I will definitely be pitching this to my fellow windows-to-linux converts.

**Comment by Fifoxtasy** [comment]
ohh, I just wanted to say thank you too!

**Comment by JeanVeupa** [comment]
That thread should stick in the top spot. A F%#!&*$!!! huge thank you.
New to linux (windows -> mint) I just regret It took me so long to find Fooyin !
Thank you so much

**Comment by twistappel** [comment]
A big thank you from me too. I've been searching for ages for something that ticks all the boxes in a music library manager.

**Comment by HongYue1** [comment]
Thank you for helping me leave windows for good! (I have always stayed for foobar2k)

**Comment by galfow** [comment]
Congratulations on getting picked up by the Arch packagers. Arch users can now get Fooyin direct from the official Extras repo.

**Comment by adri99-net** [comment]
This program is great! So well thought out. thanks !!

**Comment by Bananadurian** [comment]
❤️thanks~, i really like this software.

**Comment by RetroButton** [comment]
A little bit appreciation for the best audio player on Linux by far.
Thanks Luke for your impressive work! 😉

**Comment by ThornDuke** [comment]
A big cheer for version 0.13.0! :fireworks: 

FooYin started as a bold, revolutionary idea; by August of this year, it was already a functional and highly advanced project, and in just one month, it can already be hailed as the best music library manager (and by a wide margin!) for Linux. An application that had been sorely missed for years has now become a reality.

Congratulations and thanks to everyone involved in this project.

**Comment by johnwoltman** [comment]
I posted this topic two years ago, and I just installed 0.13.0 - Fooyin is just as amazing as ever :)

**Comment by Fifoxtasy** [comment]
Development seem really fast. If it continues like that sooner or later it will become better than foobar2000. In some ways it is already superior.
it feels pretty mature already, maybe in the next releases it could jump to 1.0 ?
